### PR TITLE
Fix CPU hist init for sparse dataset.

### LINKED ISSUE #2326: Fast histogram algorithm exhibits long start-up time and high memory usage
Two issues in one:

* Fast histogram xgboost does not start training after a while (stuck in building matrix?)
* Fast histogram xgboost uses a massive amount of memory (reaching over 50GB RAM when exact xgboost takes only a peak of about 10GB RAM)

Before running, to know:

* Will not work using Microsoft R versions (incompatible due to Microsoft R spamming messages about sending data collection to Microsoft while compiling packages)
* Dataset download is 240MB, uncompressed is 2GB, sum of RDS files are 650MB
* Make sure to have about 25GB RAM free to fast histogram (25GB extra seem committed in memory but never used for unknown reasons)
* Creating the dataset (3GB, 7,744,201,107,060 elements) requires using my svmlight package in R
* My scripts can run copy&paste just by changing the folders appropriately in `setwd`
* I expected fast histogram to be slower than exact xgboost due to sparsity, but not by that much

Created files info:

```
  File: url_svmlight.tar.gz
CRC-32: c152f632
   MD4: b05d0a58ad6f53f9ad20a5f350c25df2
   MD5: 2eb74f59d06807a3845f90ed36fe8ded
 SHA-1: 605121dde12ce7baf098e30cd8856ef4ed6d5e69
```

```
  File: reput_sparse.rds
CRC-32: 846b1907
   MD4: 2ea36b208118b7d46d18ccb5fad8da98
   MD5: ce371f418463ebc799367ed41f5fd654
 SHA-1: 7f2a73609d72ff61a24a6d176c44752b68954ea6
```

## Environment info
Operating System: Windows Server 2012 R2 (baremetal) - OS doesn't matter, happens also in Ubuntu 17.04

Compiler: MinGW 7.1

Package used (python/R/jvm/C++): R 3.4

`xgboost` version used: from source, latest commit of today (e5e7217)

## Computer specs:
* Quad E7-88xx v4 (72 cores), but 3 CPUs were disabled to get rid of NUMA
* 1TB RAM 1866MHz RAM (256GB RAM available to get rid of NUMA)
* Not virtualized

Used 8 threads for xgboost because it was for a benchmark comparison.

## Steps to reproduce

1. Download this dataset: http://archive.ics.uci.edu/ml/datasets/URL+Reputation (svmlight format, 121 parts, 2396130 rows, 3231961 features
2. Load data using my svmlight parser
3. Train a xgboost model with fast histogram method (default is 255 bins, one can start directly with 15 bins)

## What have you tried?

1. Changing xgboost to exact: starts training immediately
2. Lower number of threads used to 4: still will not start after 30 minutes
3. Lowering number of bins to 15: still will not start after 30 minutes
4. Lowering number of bins and setting number of threads to 1 (using a small laptop with same environment): eventually starts training after 5 minutes

Scripts:

LIBRARY DOWNLOAD:

```r
install.packages("devtools")
install.packages("Matrix")
install.packages("Rcpp")
install.packages("RcppEigen")
devtools::install_github("Laurae2/sparsity")
# Install xgboost separately from source
```

DATA LOAD:

```r
# Libraries
library(sparsity)
library(Matrix)

# SET YOUR WORKING DIRECTORY
setwd("E:/benchmark_lot/data")

data <- read.svmlight(paste0("Day0.svm"))
data$matrix@Dim[2] <- 3231962L
data$matrix@p[length(data$matrix@p):3231963] <- data$matrix@p[length(data$matrix@p)]
data$matrix <- data$matrix[1:(data$matrix@Dim[1] - 1), ]
label <- (data$labels[1:(data$matrix@Dim[1])] + 1) / 2
data <- data$matrix

new_data <- list()

for (i in 1:120) {
  indexed <- (i %% 10) + (10 * ((i %% 10) == 0))
  new_data[[indexed]] <- read.svmlight(paste0("Day", i, ".svm"))
  new_data[[indexed]]$matrix@Dim[2] <- 3231962L
  new_data[[indexed]]$matrix@p[length(new_data[[indexed]]$matrix@p):3231963] <- new_data[[indexed]]$matrix@p[length(new_data[[indexed]]$matrix@p)]
  new_data[[indexed]]$matrix <- new_data[[indexed]]$matrix[1:(new_data[[indexed]]$matrix@Dim[1] - 1), ]
  label <- c(label, (new_data[[indexed]]$labels[1:(new_data[[indexed]]$matrix@Dim[1])] + 1) / 2)
  
  if ((i %% 10) == 0) {
    
    data <- rbind(data, new_data[[1]]$matrix, new_data[[2]]$matrix, new_data[[3]]$matrix, new_data[[4]]$matrix, new_data[[5]]$matrix, new_data[[6]]$matrix, new_data[[7]]$matrix, new_data[[8]]$matrix, new_data[[9]]$matrix, new_data[[10]]$matrix)
    gc(verbose = FALSE)
    
    cat("Parsed element 'Day", i, ".svm'. Sparsity: ", sprintf("%05.0f", as.numeric(data@Dim[1]) * as.numeric(data@Dim[2]) / length(data@i)), ":1. Balance: ", sprintf("%04.02f", length(label) / sum(label)), ":1.\n", sep = "")
    
  }
  
}

# Save to RDS
gc()
saveRDS(data, file = "reput_sparse.rds", compress = TRUE)

# Save labels
saveRDS(label, file = "reput_label.rds", compress = TRUE)
```

RUN FAST HISTOGRAM, will take forever(?) to start:

```r
# SET YOUR WORKING DIRECTORY
setwd("E:/benchmark_lot/data")

my_data <- readRDS("reput_sparse.rds")
label <- readRDS("reput_label.rds")

library(xgboost)
data <- xgb.DMatrix(data = my_data, label = label)

gc(verbose = FALSE)
set.seed(11111)
model <- xgb.train(params = list(nthread = 8,
                                 max_depth = 5,
                                 tree_method = "hist",
                                 grow_policy = "depthwise",
                                 eta = 0.10,
                                 max_bin = 15,
                                 eval_metric = "auc",
                                 debug_verbose = 1),
                   data = data,
                   nrounds = 100,
                   watchlist = list(train = data),
                   verbose = 2,
                   early_stopping_rounds = 50)
```

RUN EXACT, will start immediately, and very fast. 10GB peak:

```r
# SET YOUR WORKING DIRECTORY
setwd("E:/benchmark_lot/data")

my_data <- readRDS("reput_sparse.rds")
label <- readRDS("reput_label.rds")

library(xgboost)
data <- xgb.DMatrix(data = my_data, label = label)

gc(verbose = FALSE)
set.seed(11111)
model <- xgb.train(params = list(nthread = 8,
                                 max_depth = 5,
                                 tree_method = "exact",
                                 eta = 0.10,
                                 eval_metric = "auc"),
                   data = data,
                   nrounds = 100,
                   watchlist = list(train = data),
                   verbose = 2,
                   early_stopping_rounds = 50)
```

ping @hcho3

### LINKED ISSUE #4503: Improve sparse dataset memory usage for hist tree method.
@hcho3  @RAMitchell 

As part of refactoring DMatrix,  I'm trying to profile memory usage issue across XGBoost.  One glaring issue in hist method is that it uses feature index to build the histogram cut matrix.  For dense dataset this is optimal, but for sparse dataset this results in very huge waste of memory.  One particular reproducible dataset would be [URL](http://archive.ics.uci.edu/ml/datasets/URL+Reputation)

The linked dataset has total size of 2.2 GB in text form.  After some optimizations and limiting to first 450000 lines of dataset, it still costs 28.6 GB of memory.  Here is a simple report from valgrind:

Screenshot from 2019-05-26 21-14-27

In some selected rows of above dataset, largest feature index is 3231962, but actual number of none zero columns is just 394.  Is there any way we can compress the histogram cut building into a form that doesn't depend on feature index?

### PR DESCRIPTION
**Pull request by trivialfis**
* Implement sparse histogram cut.
* Allow empty features.

## Result
Logging of Generating gmat from quantile hist (including indices matrix and column matrix initialization):

|     | HIGGS  (dense)     | URL (sparse)    |
|----|-------------------------|---------------------|
| Old | 36.8106 sec | +∞          |
| New | 27.9178 sec | 21.7657 sec |

I'm sure there are plenty of optimizations one can apply, but as GPU side is my priority now, those might be someone else's future work. :-)  Anyway,  I consider both #2326 and #4503 are fixed.   Sparse histogram cut is only used in non-distributed environment,  I think @hcho3 wrote a paper dealing with distributed environment, I won't get into it for now.

Closes #2326, closes #4503 .

### CODE REVIEW

**Comment by trivialfis** [comment]
Profiling GPU Hist,  maybe I can port the optimization to GPU later.

**Comment by trivialfis** [comment]
My original idea was to use CSC instead of CSR, or maybe even only storing indices for histogram algorithm.  But as pointed out by @RAMitchell there are lot more to consider when removing CSR.  So this ended up with two slightly duplicated implementations, one for sparse another for dense.  The threshold is hard coded for now and the sparse implementation is only used in rare cases, like the URL dataset.

Ideally we would still like to stuff the indices into DMatrix, and maybe support for dense internal representation.  But let's don't rush into it right now.

File: include/xgboost/data.h [hunk 1]
```diff
@@ -437,6 +437,8 @@ class DMatrix {
                        bool load_row_split,
                        const std::string& file_format = "auto",
                        const size_t page_size = kPageSize);
+
+  virtual char const* Name() const = 0;
```
Comment (RAMitchell) [inline include/xgboost/data.h]: I don't see where this is used.

**Comment by trivialfis** [comment]
I agree with the fragmented code.  My original plan was to use the sparse implementation as the only one implementation.  But calling get column batches causes significant memory usage... If hist method can reuse  the CSC matrix through out than it won't be a problem since the updater also creates something similar, but that would be a huge change to hist and I'm not ready for that.  As for lazy memory allocation, thatZ's definitely a doable option, but could you share some progress about sketching on your end so that I can avoid creating conflicting work?  I would also like to start looking at device sketching.

**Comment by RAMitchell** [comment]
I am looking at different algorithms for sketching quantiles other than the GK based algorithm used so far, so it will be a different code base for the GPU. I am currently experimenting with the count sketch algorithm with range queries.

**Comment by thvasilo** [comment]
For our work we used the [KLL sketch](https://arxiv.org/abs/1603.05346) which is AFAIK the SOTA in terms of space usage ([C++ code here](https://github.com/apache/incubator-datasketches-cpp)). The original sketch however does not support weighted quantiles, but the authors mentioned they had a weighted sketch in development/ready.

The main issue as far as we encountered with quantile sketches is that because of MPI/Rabit limitations, the byte size of the sketches needs to be known in advance. In a probabilistic sketch like this one, the only way to do that is to get the maximum possible size of the sketch and communicate that. Because we used the Parameter Server, that limitation was lifted, so we were able to only communicate the actual size of the sketch. The differences in size are several orders of magnitude:

image

In the figure above, both methods use the KLL sketch, but Block only communicates the actual size, while Row communicates the max size, as XGBoost does currently.

The same holds true for the gradient histograms.

**Comment by RAMitchell** [comment]
Thanks @thvasilo. The challenge for us will be GPU implementation. A hash table-based solution is appealing here as many independent threads can operate on this data structure in parallel. Unfortunately the space guarantees are significantly worse without modification.

File: tests/cpp/common/test_hist_util.cc [hunk 2]
```diff
@@ -34,18 +26,97 @@ TEST(HistCutMatrix, SearchGroupInd) {
...
+TEST(SparseCuts, MultiThreadedBuild) {
+  size_t constexpr kRows = 17;
+  size_t constexpr kCols = 15;
+  size_t constexpr kBins = 255;
+
+  auto Compare =
+#if defined(_MSC_VER)  // msvc fails to capture
+      [kBins](DMatrix* p_fmat) {
+#else
+      [](DMatrix* p_fmat) {
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply tests/cpp/common/test_hist_util.cc]: Clang generates a warning that says it's not necessary.  Not harmful but rather annoying. It shows up in my editor.

File: tests/cpp/common/test_hist_util.cc [hunk 3]
```diff
@@ -34,18 +26,97 @@ TEST(HistCutMatrix, SearchGroupInd) {
...
+
+  ASSERT_EQ(hmat.cut.Ptrs().size(), cuts.Ptrs().size());
+  ASSERT_EQ(hmat.cut.Ptrs(), cuts.Ptrs());
+  ASSERT_EQ(hmat.cut.Values(), cuts.Values());
+  ASSERT_EQ(hmat.cut.MinValues(), cuts.MinValues());
+
+  delete pp_mat;
+}
+
+TEST(SparseCuts, MultiThreadedBuild) {
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply tests/cpp/common/test_hist_util.cc]: See the logic in build.  I have single thread build and multi thread build, the later calls the former.

**Comment by trivialfis** [comment]
That would be correct, it's actually faster than dense implementation, most of the time are spent in get column batches and column matrix initialization.  I too share the concern of fragmented code.  So if you plan to merge your work from said paper into XGBoost, or like to use CSC through out hist method, then feel free to close this one.  You can always pick whatever useful in this PR later in your work.

**Comment by trivialfis** [comment]
> So if you plan to merge your work from said paper into XGBoost, or like to use CSC through out hist method, then feel free to close this one. You can always pick whatever useful in this PR later in your work.

@hcho3 WDYT?  If so then I won't proceed with this PR.

**Comment by hcho3** [comment]
@trivialfis Can we proceed this PR for now? It will take some time before anything from the paper gets merged here, since a parameter server like dmlc/ps-lite are quite painful to integrate into this project. 

I will create a new issue to track the long term work of enabling sparse communication in distributed setting.

**Comment by trivialfis** [comment]
@hcho3 Got it.  Thanks.

**Comment by hcho3** [comment]
@trivialfis I added the special logic for binary/ordinal values in `SparseCuts::Build()`, so that it behaves consistently as `DenseCuts::Build()`. Can you review it?

**Comment by trivialfis** [comment]
@hcho3 Thanks.  Will follow up today.

**Comment by trivialfis** [comment]
ping @hcho3  I merged the two specialization of categorical data, and addressed other comments.  Please review again. ;-)

File: src/common/hist_util.cc [hunk 4]
```diff
@@ -25,25 +25,212 @@
...
+                                   uint32_t max_num_bins,
+                                   bool const use_group_ind,
+                                   uint32_t beg_col, uint32_t end_col,
+                                   uint32_t thread_id) {
+  using WXQSketch = common::WXQuantileSketch<bst_float, bst_float>;
+  constexpr float kFactor = 8;
+
+  // Data groups, used in ranking.
+  std::vector<bst_uint> const& group_ptr = info.group_ptr_;
+  p_cuts_->min_vals_.resize(end_col - beg_col, 0);
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/common/hist_util.cc]: end_col can equal to beg_col for empty threads.  I added a `CHECK_GE`.

**Comment by hcho3** [comment]
@RAMitchell I’m inclined to merge this, now that we deduplicated AddCutPoint.
